Updates for gfortran-10.1.0 support for gentoo and mac

markjolah opened the pull request:
This is a comprehensive update, which includes several features and new packages.
* Cleanup build scripts and module templates
  * Fix `MAKE_VERBOSE` to work for relevant build scripts
  * Add `initialize_prefix_core()`, `initialize_prefix_compiler()`, and `initialize_prefix_mpi()` functions to reduce code duplication for common sections of code setting the modules and prefix for each library.
  * Remove duplicated/stale code
  * Append, don't overwrite user settings for `FFLAGS`, `CFLAGS`, `CXXFLAGS`, `LDFLAGS` in build scripts
  * Use `JEDI_OPT` environment variable
  * Remove tabs/whitespace errors
  * Conditionally add optional modules using `try-load`
  * Export `<pkgname>_ROOT` variables in package modules
* Remove build scripts no longer used
  * odb
  * oddssey
* Add build scripts for GEOS (not the model), PROJ and SQLite.
  * These are required to install Cartopy which is necessary on Orion and possibly other systems.
* Update the eckit/fckit/atlas versions to JCSDA/release-stable
* Update Mac to support gfortran-10.1.0 which is now the default in brew
  * Update LAPACK 3.7.0 -> 3.9.0 to fix compile problems
  * Check for gfortran-10 in build_mpi.sh and set `-fallow-argument-mismatch` to allow building MPICH
  * Dynamically set `NUM_PROCS`
* Update Gentoo target
  * Support for Intel compilers by adding `-gcc-name=gcc-9.3.0` to set compiler and handle libstdc++ errors under icpc

Apologies for the monolithic update, but I had to merge together all the changes from several different branches to get it working on my test systems.  So far, I have tested on:
* macOS-10.15.6 using the newest brew formula and `GFortran-10.1.0`.
* gentoo linux using a `GCC-10.1.0` base system and intel 2019.1 compiler modules piggybacking on `GCC-9.3.0` compilers for an `icpc` compatible `libstdc++`.

rickgrubin commented on README.md:
```diff
@@ -246,6 +246,36 @@ module load jedi/clang-openmpi      # set environment for subsequent JEDI builds
+
+To setup environment:
+~~~~~~~~~
+$ export JEDI_OPT=$HOME/opt/modules
```
export JEDI_OPT="${OPT:-"$HOME/opt/modules"}"

A common idiom for jedi-stack modules is to set OPT to /some/path/modules

markjolah replied:
In this case this is more of an example setup where the path would be hard-coded to wherever the user wants on that system.  It's not a script that would be generically adapting to different variables which is normally where I'm using that alternate approach.  I will make the documentation clearer.

aerorahul replied:
`OPT` should go. There should only be a `JEDI_OPT`.

mmiesch replied:
That's fine but can we keep `OPT` for now as an option?  If we want to remove it properly, the first step would be to update the JEDI documentation to use `JEDI_OPT`.  Otherwise, we will invalidate our build instructions on the various systems.

aerorahul replied:
I agree @mmiesch That should be done in a PR of its own.
There are way too many things in this one PR anyways.

rickgrubin commented on buildscripts/build_stack.sh:
```diff
@@ -85,13 +85,13 @@ build_lib CGAL cgal 5.0.2
-build_lib HDF5 hdf5 1.10.6
+build_lib HDF5 hdf5 1.12.0
 build_lib PNETCDF pnetcdf 1.12.1
 build_lib NETCDF netcdf 4.7.3 4.5.2 4.3.0
```
netCDF 4.7.4 is now considered stable; perhaps we ought to consider using it

rickgrubin replied:
Ignore per conversation in team meeting July 17 re: bugs in C++ interface for v4.7.4

mmiesch replied:
[Here's the previous PR I mentioned this morning](https://github.com/JCSDA/jedi-stack/pull/96).   I believe I was able to get 4.7.4 to work for ufo-bundle if I stuck with 4.3.0 of the C++ library.  But, @markjolah was also reporting problems with PIO so we should check that.

rickgrubin commented on buildscripts/config/config_gentoo.sh:
```diff
@@ -58,16 +58,17 @@ export       STACK_BUILD_ZLIB=N
-export       STACK_BUILD_HDF5=Y
-export    STACK_BUILD_PNETCDF=Y
-export     STACK_BUILD_NETCDF=Y
+# export    STACK_BUILD_BUFRLIB=Y
+# export       STACK_BUILD_HDF5=Y
+# export    STACK_BUILD_PNETCDF=Y
+# export     STACK_BUILD_NETCDF=Y
```
Why are these commented out? Can they not just be set to 'N' ?

markjolah replied:
DOH!  Debugging comments escaped.

rickgrubin commented on buildscripts/config/config_mac.sh:
```diff
@@ -16,12 +16,12 @@ export COMPILER_BUILD="native-pkg"
 export LOGDIR=buildscripts/log
 export OVERWRITE=Y
-export NTHREADS=4
+export NTHREADS=$(sysctl -n hw.logicalcpu)
```
export NTHREADS=${/usr/sbin/sysctl -n hw.logicalcpu}

Ensure 'sysctl' is found if not in user's PATH

mmiesch replied:
On ubuntu 18.04 it's in `/sbin`:
```bash
ubuntu@ip-172-31-72-78:~$ which sysctl
/sbin/sysctl
```

rickgrubin commented on buildscripts/libs/build_atlas.sh:
```diff
@@ -11,15 +11,14 @@ version=$2
 compiler=$(echo $JEDI_COMPILER | sed 's/\//-/g')
 mpi=$(echo $JEDI_MPI | sed 's/\//-/g')
 
-[[ $USE_SUDO =~ [yYtT] ]] && export SUDO="sudo" || unset SUDO
```
Isn't this needed for the line below:

[[ -d build ]] && $SUDO rm -rf build

mmiesch commented:
Thanks @markjolah - I'll go through this more closely and check it when I get a chance but first one request.  I do not want to use the release-stable branches of eckit, ecbuild, fckit and atlas in the stack.   If you are on an HPC system or in a container and you want to know what version of the forks is installed, the module name of `release-stable` tells you nothing.  I would like to continue using tags for the environment modules and containers.   The release-stable branches are intended for the bundles.

markjolah commented:
When building using GFortran-10 from the latets brew formula, you will run into some issues with the GFortran fixes I've already got committed to various repositories causing errors from clang like "unrecognised compiler flags `-fallow-argument-mismatch`".  These flags are ment for GFortran, but because we are mixing in the clang compilers and they don't understand this flag it's causing problems.  I am fixing these issues as they arrise.  

* First PR to fix this issue to FMS: [FMS # 40](https://github.com/JCSDA/FMS/pull/20).  Use branch `bugfix/clang-gfortran10` to test.

aerorahul commented on buildscripts/libs/build_esmf.sh:
```diff
@@ -45,9 +45,9 @@ else
-export FFLAGS="-fPIC"
-export CFLAGS="-fPIC"
-export CXXFLAGS="-fPIC"
+export FFLAGS+=" -fPIC"
```
Please don't do `+=`, here or anywhere.
These scripts are supposed to be self-contained, otherwise, we would have defined `-fPIC` in the config files long ago.

markjolah replied:
Actually, I cannot use any script that don't use the `+=` for my purposes.  I must set `-gcc-name=gcc-9.3.0` for intel compilers to find a compatible GCC version.  Without the `+= ` you are overriding my module supplied `CXXFLAGS`.  I don't see a compelling reason not to allow essential but also system-specific compiler flags to be supplied.

aerorahul commented on buildscripts/libs/build_fftw.sh:
```diff
@@ -27,10 +27,10 @@ else
-export FFLAGS="-fPIC"
-export CFLAGS="-fPIC"
+
+export FFLAGS+=" -fPIC"
```
please remove `+=`

aerorahul commented on buildscripts/libs/build_hdf5.sh:
```diff
@@ -40,9 +40,9 @@ else
-export FFLAGS="-fPIC -w"
-export CFLAGS="-fPIC -w"
-export CXXFLAGS="-fPIC -w"
+export FFLAGS+=" -fPIC -w"
```
please remove `+=`

aerorahul commented on buildscripts/libs/build_jasper.sh:
```diff
@@ -19,8 +19,8 @@ export CC=$SERIAL_CC
 export F77=$FC
-export FFLAGS="-fPIC"
-export CFLAGS="-fPIC"
+export FFLAGS+="-fPIC"
```
please remove `+=`

aerorahul commented on buildscripts/libs/build_jpeg.sh:
```diff
@@ -27,7 +27,7 @@ else
 
 export CC=$SERIAL_CC
-export CFLAGS="-fPIC"
+export CFLAGS+=" -fPIC"
```
please remove `+=`

aerorahul commented on buildscripts/libs/build_lapack.sh:
```diff
@@ -30,15 +30,17 @@ export FC=$SERIAL_FC
-export CFLAGS="-fPIC"
-export CXXFLAGS="-fPIC"
-export FCFLAGS="-fPIC"
+export FFLAGS+=" -fPIC"
```
please remove `+=`

aerorahul commented on buildscripts/libs/build_mpi.sh:
```diff
@@ -27,9 +28,11 @@ export CC=$SERIAL_CC
-export CXXFLAGS="-fPIC"
-export FCFLAGS="-fPIC"
+[[ -n $FC_GFORTRAN_10 ]] && export FFLAGS+=" -fallow-argument-mismatch"
+export FFLAGS+=" -fPIC"
```
please remove `+=`

aerorahul commented on buildscripts/libs/build_nccmp.sh:
```diff
@@ -41,8 +41,8 @@ else
 
-export CFLAGS="-fPIC"
-export LDFLAGS="-L$NETCDF_ROOT/lib -L$HDF5_ROOT/lib -L$SZIP_ROOT/lib"
+export CFLAGS+=" -fPIC"
```
please remove `+=`

aerorahul commented on buildscripts/libs/build_nccmp.sh:
```diff
@@ -41,8 +41,8 @@ else
-export CFLAGS="-fPIC"
-export LDFLAGS="-L$NETCDF_ROOT/lib -L$HDF5_ROOT/lib -L$SZIP_ROOT/lib"
+export CFLAGS+=" -fPIC"
+export LDFLAGS+=" -L$NETCDF_ROOT/lib -L$HDF5_ROOT/lib -L$SZIP_ROOT/lib"
```
please remove `+=`

aerorahul commented on buildscripts/libs/build_nco.sh:
```diff
@@ -40,15 +40,14 @@ else
-
 export F77=$FC
+
+export FFLAGS+=" -fPIC"
```
please remove `+=`

aerorahul commented on buildscripts/libs/build_nco.sh:
```diff
@@ -40,15 +40,14 @@ else
 export FCFLAGS=$FFLAGS
 
-export LDFLAGS="-L$NETCDF_ROOT/lib -L$HDF5_ROOT/lib -L$SZIP_ROOT/lib"
+export LDFLAGS+=" -L$NETCDF_ROOT/lib -L$HDF5_ROOT/lib -L$SZIP_ROOT/lib"
```
please remove `+=`

aerorahul commented on buildscripts/libs/build_netcdf.sh:
```diff
@@ -44,17 +44,17 @@ fi
-export CFLAGS="-fPIC"
-export CXXFLAGS="-fPIC -std=c++11"
-export FCFLAGS="$FFLAGS"
+export FFLAGS+=" -fPIC"
```
please remove `+=`

aerorahul commented on buildscripts/libs/build_pio.sh:
```diff
@@ -32,12 +32,12 @@ fi
-export CXXFLAGS="-fPIC"
-export FCFLAGS="$FFLAGS"
+
+export FFLAGS+=" -fPIC"
```
please remove `+=`

aerorahul commented on buildscripts/libs/build_pnetcdf.sh:
```diff
@@ -30,12 +30,12 @@ fi
-export CXXFLAGS="-fPIC"
-export FCFLAGS="$FFLAGS"
+
+export FFLAGS+=" -fPIC -w"
```
please remove `+=`

aerorahul commented on buildscripts/libs/build_szip.sh:
```diff
@@ -30,9 +30,10 @@ export CC=$SERIAL_CC
-export CFLAGS="-fPIC"
-export CXXFLAGS="-fPIC"
-export FCFLAGS="-fPIC"
+export FFLAGS+=" -fPIC"
```
please remove `+=`

aerorahul commented on buildscripts/libs/build_udunits.sh:
```diff
@@ -28,8 +28,9 @@ export FC=$SERIAL_FC
 
-export FCFLAGS="-fPIC"
-export CFLAGS="-fPIC"
+export FFLAGS+=" -fPIC"
```
please remove `+=`

aerorahul commented on buildscripts/libs/build_xerces.sh:
```diff
@@ -29,9 +29,10 @@ export FC=$SERIAL_FC
-export CFLAGS="-fPIC"
-export CXXFLAGS="-fPIC"
-export FCFLAGS="-fPIC"
+export FFLAGS+=" -fPIC"
```
please remove `+=`

aerorahul commented on buildscripts/libs/build_zlib.sh:
```diff
@@ -27,9 +27,10 @@ export FC=$SERIAL_FC
-export FCFLAGS="-fPIC"
-export CFLAGS="-fPIC"
-export CXXFLAGS="-fPIC"
+export FFLAGS+=" -fPIC"
```
please remove `+=`

aerorahul commented on buildscripts/libs/update_modules.sh:
```diff
@@ -116,6 +116,64 @@ function build_lib() {
+                                       || ( echo "WARNING: $prefix EXISTS, SKIPPING"; exit 1 )
+        fi
+    else
+        prefix=${GEOS_ROOT:-"/usr/local"}
```
What is GEOS doing here?

markjolah replied:
Copy-paste error.  Will fix.

aerorahul commented on modulefiles/mpi/compilerName/compilerVersion/mpiName/mpiVersion/atlas/atlas.lua:
```diff
@@ -19,7 +19,7 @@ always_load("fckit")
 prereq("fckit")
 
-local opt = os.getenv("OPT") or "/opt/modules"
+local opt = os.getenv("JEDI_OPT") or os.getenv("OPT") or "/opt/modules"
```
Can we get rid of `OPT` and just keep `JEDI_OPT`.

markjolah replied:
That is beyond my preview.   I introduced the JEDI_OPT because I have a conflict on some systems with that name.  I prefer JEDI_OPT, but many of the containers may still rely on that so I am keeping backwards compatibility.

mmiesch replied:
Please don't eliminate OPT right now.  Not only would we invalidate our usage instructions on ~ 7 different systems but we would also break all the modules, pulling the rug out from our users.

mmiesch replied:
It's not the containers - they don't care about OPT and they are rebuilt often enough that it's not really an issue.  But, many of the environment modules in ~15 or so different stacks that we maintain (~ 7 HPC systems, 1-3 stacks on each system) were built before `JEDI_OPT` was introduced.  So, we'd have to either rebuid all those stacks or go in and manually edit the modulefiles.  That's ok to do eventually but it will take planning and testing before implementation to make sure we don't break things.

aerorahul commented on modulefiles/mpi/compilerName/compilerVersion/mpiName/mpiVersion/eckit/eckit.lua:
```diff
@@ -14,9 +14,10 @@ local compNameVerD = compNameVer:gsub("/","-")
+try_load("eigen")
 
-prereq("netcdf","boost-headers","eigen")
+prereq("netcdf")
```
since when is `NetCDF` a prereq of `eckit`?

markjolah replied:
Good point.  That does seem to be unnecessary.

aerorahul commented on modulefiles/system/gentoo/intel/20.0.lua:
```diff
@@ -13,6 +13,7 @@ prepend_path("INTEL_LICENSE_FILE", intel_license_path)
 
 -- Compiler
 local intel_compiler_path = intel_release_path
+local intel_gcc_version = "9.3.0" -- Version of GCC to use (10.1.0) not supported yet
```
What is intel_gcc_version? some kind of oxymoron?

markjolah replied:
The intel compilers do not supply an implementation of the C++ standard library.  Rather they are designed to  piggyback on the system's GCC installation to find `libstdc++.so`.  Unfortunately, this means the intel compiler used must be paired with an appropriate version of GCC in order to actually compile C++ code.  The intel compilers are designed to take compiler flags `-gcc-name=...` arguments to set the corresponding gcc version to use.  This allows the most recent working GCC to be paired with each version of the intel compilers.

This relates to my need to pass `-gcc-name=gcc-9.3.0`.  The intel 2020.x compilers are not yet compatible with the `libstdc++.so` supplied by `GCC-10.1.0`.  must piggyback on GCC for the C++ standard library.

aerorahul replied:
Yes, OK. I understand that.  I was just confused by the choice of name for that variable. simple gcc_version would have sufficed, since it is local.

srherbener commented:
What do others think about my comment on fckit and atlas in the config_mac.sh file? Should these be "Y" or "N". My vote is "N" since I'm concerned that these are still changing often enough that one may miss necessary updates. If others feel this is not a valid concern, I'm okay leaving them set to "Y". Thanks!

aerorahul commented:
> What do others think about my comment on fckit and atlas in the config_mac.sh file? Should these be "Y" or "N". My vote is "N" since I'm concerned that these are still changing often enough that one may miss necessary updates. If others feel this is not a valid concern, I'm okay leaving them set to "Y". Thanks!

`N` for me. They have been unchanged for a while, but have a potential for changing rapidly and often.  The pain to compile in a bundle is not that great. Although it would be good to be able to disable Atlas's tests when in a bundle.

markjolah commented on buildscripts/config/config_mac.sh:
```diff
@@ -37,17 +37,17 @@ export       STACK_BUILD_ZLIB=Y
-export      STACK_BUILD_FCKIT=N
-export      STACK_BUILD_ATLAS=N
+export      STACK_BUILD_FCKIT=Y
+export      STACK_BUILD_ATLAS=Y
```
No sorry, those were again my testing changes that escaped.

markjolah commented:
@aerorahul Please see my sub comment on how the intel compilers work regarding gcc versions.  Alternative solutions are welcome, but we are perhaps stretching what is possible for a unified build script, obviously different platforms will require different compiler flags.  Also consider problems like what I encountered in the `build_mpi.sh`.  When building mpich with gfortran>=10.1.0, one needs to apply compiler flags which are specific to that compiler and version.  I consider that grep based solution hackish, but I have no alternative mechanism under jedi-stack to supply compiler flags.

Finally, consider that I would also like to pass `-g` flags to the stack in order to debug.  If each build script overrides my customized compiler flags, I have no way to do this effectively.

markjolah commented:
Regarding the need to add JEDI_OPT, many systems use OPT for other purposes which severely limits portability.  The need to add `JEDI_` directly relates to the flat namespace provided by the environment.   For maximum portability complex shell-based software needs to be careful about namespacing for environment variables that are exported.

mmiesch commented:
> @aerorahul Please see my sub comment on how the intel compilers work regarding gcc versions. Alternative solutions are welcome, but we are perhaps stretching what is possible for a unified build script, obviously different platforms will require different compiler flags. Also consider problems like what I encountered in the `build_mpi.sh`. When building mpich with gfortran>=10.1.0, one needs to apply compiler flags which are specific to that compiler and version. I consider my grep based solution there hackish, but I have no alternative mechanism under jedi-stack to supply platform-customized compiler flags.
> 
> Finally, consider that I would also like to pass `-g` flags to the stack in order to debug. If each build script overrides my customized compiler flags, I have no way to do this effectively.

I vote for taking out the `+=`, as Rahul has advised.  This is useful modularity.  Under normal conditions, we will want to build environment modules and containers with optimized versions of the dependencies.  If you want to debug some particular  package, you can rebuild it yourself in a different place.  But the assumption for the user should be that the dependencies are ready to go for a production run, like the native modules installed on HPC systems.

mmiesch commented:
And I agree that the `JEDI_` prefix is useful.  It was introduced originally for Cheyenne because their system has a dedicated meaning for `COMPILER`.  I just want to keep both `OPT` and `JEDI_OPT` around for the time being so we don't break our modules and documentation.

markjolah commented:
@mmiesch Removing the `+=` will require some other mechanism to set `CXXFLAGS` otherwise the build scripts are not going to work on a system with GCC-10 and intel at the same time.  I'm stuck for any less intrusive method.  I would propose that if you are worried about users setting flags, you could set as part of the `config_system.sh`, 
```
FFLAGS=""
CFLAGS=""
CXXFLAGS=""
```
This will give exactly the behaviour that you desire, but also allow the modules to provide `CXXFLAGS` and not be overridden by the build script.

markjolah commented:
Here's what I fixed so far.  I still need advice on how to apply system-dependent compiler flags, but I think this was all other discussed issues.

* Update NetCDF-C 4.7.3 -> 4.7.4
* Update NetCDF-Fortran 4.5.2 -> 4.5.3
* Remove debugging comments and debugging settings for fckit/atlas builds
* Correct eckit modules dependencies in build script and module file
* Check for sysctl on OSX before detecting NUM_PROCS and fall back to 4 processors
* Improve documentation for Gentoo.  Fix a few spelling errors too.
* Fix copy paste error in `_initialize_prefix()`

markjolah commented:
The only other idea I have which seemed more intrusive, is to set `JEDI_FFLAGS`, `JEDI_CFLAGS`, `JEDI_CXXFLAGS` in the `config_<system>.sh`.  Then each build script does:

```
FFLAGS="-fPIC ${JEDI_FFLAGS}"
CFLAGS="-fPIC ${JEDI_CFLAGS}"
CXXFLAGS="-fPIC ${JEDI_CXXFLAGS}"
```

And in my intel module I would need to export both:
```
CXXFLAGS="-gcc-name=<blah>"
JEDI_CXXFLAGS="-gcc-name=<blah>"
```

Would this approach be acceptable?

mmiesch commented:
> The only other idea I have which seemed more intrusive, is to set `JEDI_FFLAGS`, `JEDI_CFLAGS`, `JEDI_CXXFLAGS` in the `config_<system>.sh`. Then each build script does:
> 
> ```
> FFLAGS="-fPIC ${JEDI_FFLAGS}"
> CFLAGS="-fPIC ${JEDI_CFLAGS}"
> CXXFLAGS="-fPIC ${JEDI_CXXFLAGS}"
> ```
> 
> And in my intel module I would need to export both:
> 
> ```
> CXXFLAGS="-gcc-name=<blah>"
> JEDI_CXXFLAGS="-gcc-name=<blah>"
> ```
> 
> Would this approach be acceptable?

Thanks @markjolah - I like the modularity as I said but at the same time I do appreciate your point.  Another place I ran into this was on Discover where at least hdf5 had to be compiled with the `-m64` flag or it would fail (meanwhile I had to exclude that flag for the jedi bundles).  Also, there are situations where you want to compile with, say, `-O2` or `-O3` but this might fail for particular packages and platforms/compilers where you need to reduce the optimization.  So, some balance is needed between local control and global application.  

So, that said, I think I do like your second approach with the `JEDI_CXXFLAGS`.  This makes it unambiguous what flags are being specified.  There is the global application part and the local control.   And, it's clear what is being set.  One worry with the `+=` is you don't know for sure if you missed one of multiple places where some variable is being appended.

In the second approach, why do you need to export both in your modules?  These are only needed for the build, right, so why can't you just export `JEDI_CXXFLAGS` from the `build_stack.sh` script?

aerorahul commented:
> The only other idea I have which seemed more intrusive, is to set `JEDI_FFLAGS`, `JEDI_CFLAGS`, `JEDI_CXXFLAGS` in the `config_<system>.sh`. Then each build script does:
> 
> ```
> FFLAGS="-fPIC ${JEDI_FFLAGS}"
> CFLAGS="-fPIC ${JEDI_CFLAGS}"
> CXXFLAGS="-fPIC ${JEDI_CXXFLAGS}"
> ```
> 
> And in my intel module I would need to export both:
> 
> ```
> CXXFLAGS="-gcc-name=<blah>"
> JEDI_CXXFLAGS="-gcc-name=<blah>"
> ```
> 
> Would this approach be acceptable?

@markjolah 
This solution works. Thanks!  I am also aware of what @mmiesch had to do with discover.  I think `${JEDI_}` is a good and acceptable solution.

While we are on this topic, I want to suggest that we get into a habit of creating a branch/tag for the deployments we do on HPC's and containers, rather than just checkout `develop` and make changes willy-nilly, deploy and then trash the checkout.

The reason is the ability to track changes and reproducibility on these platforms, so that if issues arise, the stack can be reproduced and problem identified.  That's just a thought for the future.

rickgrubin commented:
> While we are on this topic, I want to suggest that we get into a habit of creating a branch/tag for the deployments we do on HPC's and containers, rather than just checkout `develop` and make changes willy-nilly, deploy and then trash the checkout.
> 
> The reason is the ability to track changes and reproducibility on these platforms, so that if issues arise, the stack can be reproduced and problem identified. That's just a thought for the future.

@aerorahul you are not the only person who feels this way. The knots on my forehead, attributable to dealing with the idiosyncrasies of each host, are a testament to that.

Along those lines, I'd prefer to see that the modulefiles directory structure be consistent; that is, always:

- `module use $OPT/modulefiles/core` 
- `module load jedi/<whatever>`

or use 'apps' instead of 'core.' Regardless, be consistent across hosts.

aerorahul commented:
> > While we are on this topic, I want to suggest that we get into a habit of creating a branch/tag for the deployments we do on HPC's and containers, rather than just checkout `develop` and make changes willy-nilly, deploy and then trash the checkout.
> > The reason is the ability to track changes and reproducibility on these platforms, so that if issues arise, the stack can be reproduced and problem identified. That's just a thought for the future.
> 
> @aerorahul you are not the only person who feels this way. The knots on my forehead, attributable to dealing with the idiosyncrasies of each host, are a testament to that.
> 
> Along those lines, I'd prefer to see that the modulefiles directory structure be consistent; that is, always:
> 
> * `module use $OPT/modulefiles/core`
> * `module load jedi/<whatever>`
> 
> or use 'apps' instead of 'core.' Regardless, be consistent across hosts.

@rickgrubin 
`core`, `compiler` and `mpi` are names suggested from `Lmod`.  I have heard the complaint about `core` being a poor choice, but oh well!  `core` is suggestive of compiler independent libraries. It is sometimes just not worth it to stray from the suggested nomenclature.

We do have `apps`, which contain meta-modules e.g. `apps/jedi/gnu-9.3.0-openmpi-4.0.1.lua` etc., which loads the modules and the hierarchy as if one would by doing every individual module load. 
In my world, I also have `apps/ufs/gnu-9.3.0-openmpi-4.0.1.lua`. So there it is. 

There are other software packaging and deployment options e.g. `spack`, that make `$$$`, doing this. It is a poor-man's solution and works (most of the time).  If this becomes a full-time need, one should seriously consider some of the "professional" solutions for packaging and deployment.

markjolah commented:
@mmiesch 

> In the second approach, why do you need to export both in your modules? These are only needed for the build, right, so why can't you just export JEDI_CXXFLAGS from the build_stack.sh script?

The export of regular `CXXFLAGS` is so I can also compile non-jedi-stack things, i.e., so the compilers work as expected with modules loaded.

RE: flags, it seems there is both a need and a consensus approach.  I'll fix this up and re-test a bit later this week.

RE: Modulefiles folders and names:  I like keeping it uniform, whatever names are OK with me since the `module use` command lets you switch things up dynamically, there are less consequences for name conflicts than with environment variables.  

I also like @aerorahul's concern about versioning and github.  Also we need a regular pattern for keeping version numbers on overall `apps`, so that each prefix can have a stable app module name and version associated with it that will not change.  This is necessary to e.g., profile older code alongside newer code, or to keep a compiled workflow working correctly whitest the maintainer is updating the stack to a new version.

mmiesch commented:
> The export of regular `CXXFLAGS` is so I can also compile non-jedi-stack things, i.e., so the compilers work as expected with modules loaded.

Yes, but my point was that you maybe don't need to include `JEDI_CXXFLAGS` in your modules.   It's only used for building the jedi-stack - it's not needed for building the bundles or anything else.  So, you could set it in the `config_*.sh` scripts or the `build_stack.sh` script to isolate it's influence.  It might be useful for others to have the capability to set those in one of the jedi-stack scripts.

mmiesch commented:
Regarding @aerorahul 's comment - yes, I agree we should be more diligent about saving our configurations for different systems on GitHub.   I have also discussed this with @rickgrubin.  There are trade-offs here, though.  For example, several systems like gadi (Australia) and now AWS use tcl modules.    There is now a feature/gadi branch maintained by @jintaglee that houses these modulefiles.  But, if we merge this in to develop, then we will be making a commitment to maintain the tcl modulefiles along with the lua modulefiles.  That will take effort to test and keep up to date and, it may not be necessary if we move more toward spack, particularly for the tcl systems.

markjolah commented:
@mmiesch 

> Yes, but my point was that you maybe don't need to include JEDI_CFLAGS in your modules. It's only used for building the jedi-stack - it's not needed for building the bundles or anything else. So, you could set it in the config_*.sh scripts or the build_stack.sh script to isolate it's influence. It might be useful for others to have the capability to set those in one of the jedi-stack scripts.

This gets to the core problem of this `JEDI_CFLAGS` approach and why I still think appending to `CFLAGS` and not overwriting it is the correct way to implement this.  Note that each build script is in charge of purging and then loading the modules that it requires including the compiler.  It is only after the compiler modules are loaded that the `CFLAGS` for the specific package can be set.  Some build scripts (the majority) do not touch the flags, for those I need `CFLAGS="-gcc-name=gcc-9.3.0"` to build.  For the select build scripts which overwrite these variables forcibly, I need to provide `JEDI_CFLAGS="-gcc-name=gcc-9.3.0"` in order for my flag to make it through to the configure step.  This step would not be necessary if the flags were not being overwritten.

I've pushed an update which reworks everything as requested by reviewers and it seems to build for me the same as before.

I would point out all other build systems I have worked with take great pains to *not* overwrite the user's `CFLAGS`, and it's informative to look at how these other systems all deal with this same problem:
* [FreeBSD ports](https://www.freebsd.org/ports/) the granddaddy of all package build-systems says [Respect CFLAGS!](https://www.freebsd.org/doc/en_US.ISO8859-1/books/porters-handbook/dads-cflags.html)
* [Gentoo portage](https://devmanual.gentoo.org/eclass-writing/index.html) (inspired by BSD ports) lets users set their `CFLAGS` at the top level `make.conf` and all *ebuilds* must respect them.  They include [facilities to filter-in or filter-out flags](https://devmanual.gentoo.org/eclass-reference/flag-o-matic.eclass/index.html) as necessary and functions to test if a flag is acceptable to a compiler.
* The [MXE cross-compiling environment](https://mxe.cc/) build scripts use Make and a quick look shows most packages avoid setting CFLAGS, therefore using the environment CFLAGS, or else they append to CFLAGS  
* Spack let's users [specifiy custom CFLAGS for each compiler](https://spack-tutorial.readthedocs.io/en/latest/tutorial_configuration.html#compiler-flags).  Then the build system is Python based, but again [the design respects the user CFLAGS if possible](https://spack.readthedocs.io/en/latest/packaging_guide.html#compiler-flags), and append operations are typically used to add package-specific flags to a *spec*.
* [Homebrew Formula](https://github.com/Homebrew/homebrew-core/tree/master/Formula) almost always use `ENV.append "CFLAGS", "-foo"` when they need to add a flag, they do not mess with the global set flags.

mmiesch commented:
Ok - maybe you gave up your argument too soon - if you mentioned all this before, maybe I could have been convinced!  But, you don't have to change it back.   I don't follow this bit though:

>Note that each build script is in charge of purging and then loading the modules that it requires 

I hope each build script does not purge the modules.  This is not the way it was done before and I do not see any new module purge commands.  This is the way it should be because some systems require you to load some common background modules prior to running any of these scripts.  Examples include the `license_intel` module on S4 and the cmake, git, and python modules on various other systems.

rickgrubin commented on buildscripts/libs/build_nccmp.sh:
```diff
@@ -41,8 +41,8 @@ else
+export CFLAGS="-fPIC ${JEDI_CFLAGS}"
+export LDFLAGS="-L$NETCDF_ROOT/lib -L$HDF5_ROOT/lib -L$SZIP_ROOT/lib ${JEDI_LDFLAGS}"
 
 url="https://gitlab.com/remikz/nccmp/-/archive/$version/${software}.tar.gz"
```
Looks like this tool's git structure has changed _archive_ to _tree_
```suggestion
url="https://gitlab.com/remikz/nccmp/-/tree/$version/${software}.tar.gz"
```

markjolah commented:
Based on discussion at the 08/07 Jedi-core meeting regarding this PR, it has been decided the cleanest approach that will satisfy everyone and also allow the possibility of building debug stacks is to use the `+=` syntax for CFLAGS as is suggested in almost every other build framework (see above references).  In order to prevent the mistaken (miss)-use of this feature each of the system config scripts will set the CFLAGS to empty.  The net effect is the system works the exact same as before, except that if loading a compiler module requires modifying your CFLAGS, then those changes will correctly take effect.

I have merged with develop and tested on `mac` and `gentoo` systems.  I also cleaned up some whitespace issues.

mmiesch commented on buildscripts/build_stack.sh:
```diff
@@ -89,13 +89,13 @@ build_lib CGAL cgal 5.0.2
-build_lib ATLAS atlas jcsda 0.20.2.jcsda1
+build_lib ECKIT eckit jcsda release-stable
+build_lib FCKIT fckit jcsda release-stable
+build_lib ATLAS atlas jcsda release-stable
```
Please change these back to the tagged versions for building modules

mmiesch commented on buildscripts/libs/build_atlas.sh:
```diff
@@ -15,15 +14,14 @@ version=$2
     module load jedi-$JEDI_COMPILER
     module load jedi-$JEDI_MPI
-    module load ecbuild eckit fckit
+    module try-load ecbuild
```
Why a try-load here?  You're not going to get far without it

markjolah replied:
I gentoo uses the same ecbuild for the system builds and the stack builds, so I never build an ecbuild module, and there is no module to load.

mmiesch commented on buildscripts/libs/build_eckit.sh:
```diff
@@ -15,19 +14,16 @@ version=$2
-    module load boost-headers eigen
-    module load ecbuild
+    module try-load boost-headers
+    module try-load eigen
```
These too - the `try-load` gives problems on tcl-based module systems.   It's ok now and then for things like eigen that are really optional but why use it for, e.g. ecbuild?  If ecbuild is not found then this script is going to fail.  Is it because you install these in some standard location, not as modules?

markjolah replied:
I also have no eigen or boost-headers modules, as my base system provides them.

mmiesch replied:
Ok - as I said, this causes problems with tcl module systems but these are easily solved by adding a `tcl_setup` script that just does a global replace of `try-load` with `load`.  The tcl modulefile heirarchy is more of a problem.

mmiesch commented on buildscripts/libs/build_bufrlib.sh:
```diff
@@ -43,7 +43,7 @@ cd ${JEDI_STACK_ROOT}/${PKGDIR:-"pkg"}
 git checkout $version
 
-VERBOSE="$MAKE_VERBOSE" $SUDO ./tools/build.sh $prefix -DBUILD_SHARED_LIBS=1
+VERBOSE=$MAKE_VERBOSE $SUDO ./tools/build.sh $prefix -DBUILD_SHARED_LIBS=1
```
Do we really need to call another script here?  This is failing on AWS because the modules are not loaded in `build.sh`.  It's using the system cmake (version 3.10) because it can't see the module cmake (version 3.16).  Also, I suspect that even if it made it past that step it would not get the correct compiler if the compiler was provided via a module.

markjolah replied:
Didn't notice this was working through the convenience script, this edit was just part of regex search-replace.  I changed to direct call to cmake.  Let me know if you still have problems.

mmiesch commented:
strange - I'm getting this error on AWS/ubuntu when compiling ufo-bundle after building this stack:
```
[ 69%] Linking CXX shared library ../../../lib/libqg.so
/usr/bin/ld: /opt/modules/gnu-9.3.0/lapack/3.9.0/lib/liblapack.a(dgetrf2.f.o): relocation R_X86_64_PC32 against symbol `dgetrf2_' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
oops/qg/model/CMakeFiles/qg.dir/build.make:1226: recipe for target 'lib/libqg.so' failed
make[2]: *** [lib/libqg.so] Error 1
CMakeFiles/Makefile2:11336: recipe for target 'oops/qg/model/CMakeFiles/qg.dir/all' failed
make[1]: *** [oops/qg/model/CMakeFiles/qg.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

I think the space is causing problems.

mmiesch commented on buildscripts/libs/build_lapack.sh:
```diff
@@ -34,15 +33,17 @@ export FC=$SERIAL_FC
-export CFLAGS="-fPIC"
-export CXXFLAGS="-fPIC"
-export FCFLAGS="-fPIC"
+export FFLAGS+=" -fPIC"
```
Yes - confirmed.  If the user does not set flags (the default), then this space in front of `-fPIC` confuses cmake and it does not add the flag.

with the space:

```
cd /home/ubuntu/jedi/jedi-stack/pkg/lapack-3.9.0/build/BLAS/SRC && /usr/bin/gfortran   -O2 -DNDEBUG -O2   -c /home/ubuntu/jedi/jedi-stack/pkg/lapack-3.9.0/BLAS/SRC/isamax.f -o CMakeFiles/blas.dir/isamax.f.o
```

without the space:
```
cd /home/ubuntu/jedi/jedi-stack/pkg/lapack-3.9.0/build/BLAS/SRC && /usr/bin/gfortran   -fPIC -O2 -DNDEBUG -O2   -c /home/ubuntu/jedi/jedi-stack/pkg/lapack-3.9.0/BLAS/SRC/sasum.f -o CMakeFiles/blas.dir/sasum.f.o
```

markjolah replied:
Interesting.  I will need to investigate this one.

mmiesch replied:
A simple fix may just be a bash statement to remove leading whitespace.  But it would have to be in all the scripts.  An alternative may be *add* a trailing whitespace (if needed) in `build_stack.sh` if the base flags set in the config files are not empty.

markjolah replied:
@mmiesch How do you do the stack build on AWS/ubuntu?  I've got a EC2 instance running the `jedi single-node stack v0.2` AMI.  I don't see a config script in jedi-stack for AWS, and there weren't any instructions on the main README.

mmiesch replied:
I just use the `custom_config.sh` file for most systems.  And the instructions are just as described on the top-level readme - no specific instructions now.

However - one (big) caveat - if you use the latest `jedinode.py` script to create your node, the modules are tcl.  So, the automated modulefile building won't work.  The build scripts will work fine but currently I have just been maintaining the modulefiles manually.  Not ideal, I know - we need a better system.  

I can set it up so that the Parallelcluster nodes use tcl while the single devel nodes use lmod but I have not yet done so. see https://github.com/JCSDA/jedi-tools/issues/91

markjolah replied:
Yes so I did run into that problem with TCL modules not understanding `try-load`.  Because of that I can't easily test it is working in that environment, but hopefully this has fixed the build.

markjolah commented:
@srherbener I'm not intending to change the versions of eckit etc, those release-stable tags escaped testing.  I'm still trying to fix that eckit SQL yacc bug, sorry that's probably the reason eckit release-stable did not work for you.

markjolah commented:
I fixed some items, but others still require some work.
Fixed:
* Use tagged versions of ecmwf packages
* Fix proj module lib directories
* Fix build script for bufrlib to use cmake directly

TODO:
[ ] - Figure out eckit SQL/yacc issue.  Possibly disable SQL support we never use.
[ ] - Solve mysterious space problem in CFLAGS/LAPACK

markjolah commented:
Hopefully this will fix the LAPACK build.  The issue is they have several CMake bugs in the 3.9.0 release.  I downgraded to 3.8.0 and played around with the cmake variables that are set, so that it is working both when extra flags are injected from the compiler module and in the normal case where the flags are empty.  I ended up moving the space to the end of the flags here.  I'm not sure if that is even still an issue for this version.  The LAPACK developers are mixing up lists and strings in CMake which is the root of these problems.  In CMake lists are separated with `;` symbols and they are manipulating the flags like lists, but the flags actually one big string separated by spaces, not a CMake list.  With the 3.9.0 version the LAPACK CMake was adding the `;` between the flags which on the command line means "end of line".

mmiesch commented:
Thanks - I was using cmake 3.16.3 so I don't think that was the problem.  I'll try again later today

mmiesch commented:
Also - LAPACK was only the first one to fail - I suspect other cmake builds would have failed similarly so maybe more changes like that are needed?

markjolah commented:
@mmiesch You misread.  The problem is not with CMake, but with LAPACK specifically.  It is broken, and we're just working around their bugs, all the other packages work OK (that I know of).

markjolah commented:
I am working through a more comprehensive build of all the packages on an older Ubuntu ec2 instance with LMod.  The LAPACK-3.8.0 seems to be working fine now.

I pushed an update of eigen-3.3.5 -> eigen-3.3.7, which was precipitated because the url links are broken.  I fixed the links and tarball names, and figured we might as well update and test this all at once.  I have been using eigen-3.3.7, which has a few minor bugfixes and the tests should pass for all the containers.

I will see if anything other urls are broken/obsolete or there are any other unintended interactions with the CFLAGS.

mmiesch commented:
This looks good to me.  I built all the main jedi packages on AWS/ubuntu/lmod and everything built and all ufo-bundle tests pass.

rickgrubin commented on buildscripts/config/config_cheyenne.sh:
```diff
@@ -70,4 +75,6 @@ export      STACK_BUILD_BASELIBS=N
 export          STACK_BUILD_TAU2=N
 export          STACK_BUILD_CGAL=N
-
+export          STACK_BUILD_GEOS=N
+export        STACK_BUILD_SQLITE=N
```
For consistency with other config scripts, add the comments below, or remove from the `config` scripts altogether:

```
export        STACK_BUILD_GEOS=Y    #Note this Geometry-engine open-source not GEOS Model
export        STACK_BUILD_SQLITE=Y  #Required for PROJ
```

If you choose to add it, the only config script I found it in is: `config_orion.sh`

srherbener commented on buildscripts/build_stack.sh:
```diff
@@ -80,18 +80,18 @@ build_lib JPEG jpeg 9.1.0
 #----------------------
 # These must be rebuilt for each MPI implementation
-build_lib HDF5 hdf5 1.10.6
+build_lib HDF5 hdf5 1.12.0
```
A discussion has come up about compatibility between hdf5 1.12.0 and h5py (JCSDA/jedi-stack#117). While there are a lot of nice features in hdf5 1.12.0 (which we would eventually like to have), there are some compatibility issues as pointed out in the release article from the HDF group:

https://www.hdfgroup.org/wp-content/uploads/2020/03/Webinar-Moving-applications-from-1.10-to-HDF5-1.12-release-.pdf

One thing to note is that files created with hdf5 1.12.0 may not be readable by applications built on 1.10.*.

How thoroughly have we tested hdf5 1.12.0 on HPC systems? I guess something that is helping us is that most (maybe all) applications are going through the netcdf API which should hide the HDF5 API changes from clients.

rhoneyager replied:
> One thing to note is that files created with hdf5 1.12.0 may not be readable by applications built on 1.10.*.

We aren't yet using any features that would break older readers. 

> I guess something that is helping us is that most (maybe all) applications are going through the netcdf API which should hide the HDF5 API changes from clients.

For any of the code that depends on netCDF: netCDF already sets the appropriate API feature macros so that it can be linked against any version of HDF5. 

For ioda-engines: the code is natively compatible with both HDF5 1.12 and 1.10's APIs.

srherbener replied:
Thanks for the information @rhoneyager.

I'm voting to forge ahead with hdf5 1.12.0 since you noted in #117 that the conflict with h5py has been resolved. I've been using hdf5 1.12.0 on my Macs for the past month (builds from this PR) without any issues. Plus it will be good to get hdf5 1.12.0 established since there are some interesting features for future ioda development.

mmiesch commented:
I'll go ahead and merge this - we can leave hdf 1.12 in for now and experiment a bit with it.  We can revert to 10.6 if there are any problems.

mmiesch commented:
Oh - @aerorahul - it looks like you have requested changes that are unresolved so it's giving me push back on the merge - are you ok with merging?
